src/forcing/ForcingsEngine.cpp

```diff
@@ -0,0 +1,239 @@
+  -> const clock_type::time_point&
+{
+    return time_end_;
+}
+
+auto ForcingsEngine::time_step() const noexcept
+  -> const clock_type::duration&
+{
+    return time_step_;
+}
+
+auto ForcingsEngine::time_index(time_t ctime) const noexcept
+  -> size_type
+{
+    const auto epoch = clock_type::from_time_t(ctime);
+
+    if (epoch < time_begin() || epoch > time_end()) {
+        return bad_index;
+    }
+
+    const auto offset = (epoch - time_begin()) / time_step();
+    return offset;
+}
+
+auto ForcingsEngine::divide_index(const std::string& divide_id) const noexcept
+  -> size_type
+{
+
+    const auto  id_sep   = divide_id.find('-');
+    const char* id_split = id_sep == std::string::npos
+                           ? divide_id.data()
+                           : &divide_id[id_sep + 1];
+    const int   id_int   = std::atoi(id_split);
+
+    const auto pos = std::lower_bound(var_divides_.begin(), var_divides_.end(), id_int);
+    if (pos == var_divides_.end() || *pos != id_int) {
+        return bad_index;
+    }
+
+    // implicit cast to unsigned
+    return std::distance(var_divides_.begin(), pos);
+}
+
+auto ForcingsEngine::variable_index(const std::string& variable) const noexcept
+  -> size_type
+{
+    const auto vars = outputs();
+    const auto* pos = std::find_if(vars.begin(), vars.end(), [&](const std::string& var) -> bool {
+        return var == variable || var == (variable + "_ELEMENT");
+    });
+    
+    if (pos == vars.end()) {
+        return bad_index;
+    }
+
+    // implicit cast to unsigned
+    return std::distance(vars.begin(), pos);
+}
+
+double ForcingsEngine::at(
+    const time_t& raw_time,
+    const std::string& divide_id,
+    const std::string& variable
+)
+{
+    const auto t_idx = time_index(raw_time);
+    const auto i_idx = divide_index(divide_id);
+    const auto v_idx = variable_index(variable);
+    
+    if (t_idx == bad_index || i_idx == bad_index || v_idx == bad_index) {
+        const auto shape = var_cache_.shape();
+        const auto shape_str = std::to_string(shape[0]) + ", " +
+                               std::to_string(shape[1]) + ", " +
+                               std::to_string(shape[2]);
+
+        const auto index_str = (t_idx == bad_index ? "?" : std::to_string(t_idx)) + ", " +
+                               (i_idx == bad_index ? "?" : std::to_string(i_idx)) + ", " +
+                               (v_idx == bad_index ? "?" : std::to_string(v_idx));
+
+        std::string time_str = std::ctime(&raw_time);
+        time_str.pop_back();
+
+        throw std::out_of_range{
+            "Failed to get ForcingsEngine value at index {" + index_str + "}" +
+            "\n  Shape    : {" + shape_str + "}" +
+            "\n  Time     : `" + time_str  + "`" +
+            "\n  Divide   : `" + divide_id + "`" +
+            "\n  Variable : `" + variable  + "`"
+        };
+    }
+
+    // If time index is past current index, 
+    // update until new index and cache values.
```

Comment:
Is there any mechanism (planned) to discard data from timesteps earlier than (some window before) the current one? Or is it intentional to accumulate the forcings values for the entire run?

Reply:
The usual meaning of a cache is something with a fixed or limited size, providing ready access to values that would otherwise be accessed from a slower source (e.g. re-read/compute from the backing data)

Reply:
For a longer run, this is ultimately going to be a big concern, since it'll be a growing memory footprint without bound.

Reply:
My bad, I thought I replied to this one 😅. Yeah I want to reduce the memory footprint this takes up by only storing the last timestep's variable values. Just haven't refactored that in yet